python/dolfinx/nls/petsc.py

```diff
@@ -72,3 +91,453 @@ def setP(self, P: types.FunctionType, Pmat: PETSc.Mat):  # type: ignore
+        residual: List of forms of the residual
+        jacobian: List of list of forms of the Jacobian
+        bcs: List of Dirichlet boundary conditions
+        snes: The solver instance
+        x: The vector containing the latest solution
+        F: Vector to assemble the residual into
+    """
+    assert x.getType() != "nest", "Vector x should be non-nested"
+    assert F.getType() != "nest", "Vector F should be non-nested"
+    with F.localForm() as f_local:
+        f_local.set(0.0)
+
+    replace_solution_block(u, x)
+    assemble_vector_block(F, residual, jacobian, bcs=bcs, x0=x, alpha=-1.0)  # type: ignore
+
+
+def J_block(
+    u: list[dolfinx.fem.Function],
+    jacobian: list[list[dolfinx.fem.Form]],
+    preconditioner: typing.Optional[list[dolfinx.fem.Form]],
+    bcs: list[dolfinx.fem.DirichletBC],
+    snes: PETSc.SNES,  # type: ignore
+    x: PETSc.Vec,  # type: ignore
+    J: PETSc.Mat,  # type: ignore
+    P: PETSc.Mat,  # type: ignore
+):
+    """Assemble the Jacobian matrix.
+
+    Args:
+        u: List of functions tied to the solution vector within the residual and jacobian
+        jacobian: List of list of forms of the Jacobian
+        preconditioner: List of forms of the preconditioner
+        bcs: List of Dirichlet boundary conditions
+        snes: The solver instance
+        x: The vector containing the latest solution
+        J: Matrix to assemble the Jacobian into
+        P: Matrix to assemble the preconditioner into
+    """
+    replace_solution_block(u, x)
+    assert x.getType() != "nest", "Vector x should be non-nested"
+    assert J.getType() != "nest", "Matrix J should be non-nested"
+    assert P.getType() != "nest", "Matrix P should be non-nested"
+    J.zeroEntries()
+    assemble_matrix_block(J, jacobian, bcs=bcs, diagonal=1.0)  # type: ignore
+    J.assemble()
+    if preconditioner is not None:
+        P.zeroEntries()
+        assemble_matrix_block(P, preconditioner, bcs=bcs, diagonal=1.0)  # type: ignore
+        P.assemble()
+
+
+def replace_solution_block(u: list[dolfinx.fem.Function], x: PETSc.Vec):  # type: ignore
+    """Update the solution for the unknown `u` with the values in `x`.
+
+    Args:
+        u: List of function to insert data into
+        x: Vector to copy data from0
+    """
+    x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)  # type: ignore
+    offset_start = 0
+    for ui in u:
+        Vi = ui.function_space
+        num_sub_dofs = Vi.dofmap.index_map.size_local * Vi.dofmap.index_map_bs
+        ui.x.petsc_vec.array_w[:num_sub_dofs] = x.array_r[
+            offset_start : offset_start + num_sub_dofs
+        ]
+        ui.x.petsc_vec.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)  # type: ignore
+        offset_start += num_sub_dofs
+
+
+def copy_solution_block(u: list[dolfinx.fem.Function], x: PETSc.Vec):  # type: ignore
```

Comment:
We should review the naming in the more general context - u doesn't need to be a solution, necessarily.